src/models/state/mod.rs

```diff
@@ -37,25 +20,57 @@ use super::blockchain::type_scripts::TypeScript;
 use super::consensus::tasm::program::ConsensusProgram;
 use super::consensus::timestamp::Timestamp;
 use crate::config_models::cli_args;
+use crate::database::storage::storage_schema::traits::StorageWriter as SW;
+use crate::database::storage::storage_vec::traits::*;
+use crate::database::storage::storage_vec::Index;
 use crate::locks::tokio as sync_tokio;
 use crate::models::peer::HandshakeData;
 use crate::models::state::wallet::monitored_utxo::MonitoredUtxo;
 use crate::models::state::wallet::utxo_notification_pool::ExpectedUtxo;
+use crate::prelude::twenty_first;
 use crate::time_fn_call_async;
 use crate::util_types::mutator_set::addition_record::AdditionRecord;
+use crate::util_types::mutator_set::commit;
 use crate::util_types::mutator_set::ms_membership_proof::MsMembershipProof;
+use crate::util_types::mutator_set::mutator_set_accumulator::MutatorSetAccumulator;
 use crate::util_types::mutator_set::removal_record::RemovalRecord;
-
 use crate::{Hash, VERSION};
+use anyhow::{bail, Result};
+use itertools::Itertools;
+use num_traits::CheckedSub;
+use std::cmp::max;
+use std::ops::{Deref, DerefMut};
+use tracing::{debug, info, warn};
+use twenty_first::math::bfield_codec::BFieldCodec;
+use twenty_first::math::digest::Digest;
+use twenty_first::util_types::algebraic_hasher::AlgebraicHasher;
 
 pub mod archival_state;
 pub mod blockchain_state;
 pub mod light_state;
 pub mod mempool;
 pub mod networking_state;
 pub mod shared;
+mod utxo_receiver;
 pub mod wallet;
 
+pub use utxo_receiver::ChangeNotifyMethod;
+pub use utxo_receiver::UtxoNotifyMethod;
+pub use utxo_receiver::UtxoReceiver;
+
+#[derive(Debug, Clone)]
+pub struct TransactionDetails {
+    pub receiver_data: Vec<UtxoReceiver>,
+    pub removal_records: Vec<RemovalRecord>,
+    pub addition_records: Vec<AdditionRecord>,
+    pub input_utxos: Vec<(Utxo, LockScript, MsMembershipProof)>,
+    pub output_utxos: Vec<Utxo>,
+    pub public_announcements: Vec<PublicAnnouncement>,
+    pub expected_utxos: Vec<ExpectedUtxo>,
+    pub fee: NeptuneCoins,
+    pub timestamp: Timestamp,
+}
```

Comment:
There seems to be a lot of overlap with `TransactionKernel` and `transaction::primitive_witness::PrimitiveWitness`. Let's a) avoid code duplication, and b) use names specific enough to meaningfully distinguish similar models.

In my mental model:
 - A `TransactionKernel` is the unit of data that appears on the blockchain and effects a transfer of tokens. It is generally circulated with a proof in which case the thing is upgraded to a `Transaction`. The `proof` field of a `Transaction` certifies the validity of the `TransactionKernel`.
 - A `PrimitiveWitness` is all the data you need to generate a transaction proof in the base case (*i.e.*, no recursion. Where the `TransactionKernel` contains commitments, the `PrimitiveWitness` contains the preimages to these commitments.

I'm not sure what the purpose of `TransactionDetails` is so I can't comment on whether it needs all the fields it has or what a better name might be. Please elaborate.

Reply:
Sure.

`TransactionDetails` started as a private struct for passing data between some of the create_transaction() worker methods, which previously just used tuples. In particular `generate_tx_data_for_transaction()` returns a `TransactionDetails` to `create_transaction()` which then passes it to `create_transaction_from_tx_data()`.

I decided to make it public and return it from `CreateTransaction` because it contains some extra data that is not available in the `Transaction` itself. 

But yes you are right there is a lot of duplication, and that is not clean, so it is good you mentioned about it.  The fields not in TransactionKernel are:

```
    pub receiver_data: Vec<UtxoReceiver>,
    pub input_utxos: Vec<(Utxo, LockScript, MsMembershipProof)>,
    pub output_utxos: Vec<Utxo>,
    pub expected_utxos: Vec<ExpectedUtxo>,
```

The only required field for present usage is `unexpected_utxos`, which is necessary for the caller to inform the wallet of these.  So at simplest we could have `create_transaction()` return (Transaction, Vec&lt;ExpectedUtxo>) instead and perhaps that is best for now.

I think `TransactionDetails` could remain, only as an implementation detail private to the module.

Reply:
question:  would it be a bad idea to add an expected_utxos field to `Transaction` itself?   Then create_transaction() could just return Transaction, as before, and that would be much cleaner.

Previously I was thinking I shouldn't touch Transaction as it goes on the wire, but you are saying that really it is TransactionKernel on the wire, so maybe it's ok?

and/or am open to any suggestion that makes the create_transaction() api cleaner.